### PR TITLE
cryptol-saw-core: Adapt to PLiteralLessThan in cryptol

### PR DESCRIPTION
**Pull request by RyanGlScott**
This was introduced in GaloisInc/cryptol@ddbb673.

### CODE REVIEW

**Comment by robdockins** [comment]
Thanks for looking into this.  I don't think this is really the right fix, unless we view this just as a band-aid to fix the compile.

Similar to `PLiteral`, this class has some computational content we need to capture, and we should also implement the associated value-level operation, `fromToLessThan`.

**Comment by RyanGlScott** [comment]
Thanks! Admittedly, I only implemented the smallest possible change needed to make GaloisInc/saw-script#1137 go through. Unfortunately, I now realize that that causes SAW to fail at runtime, so I really should take the opportunity to do this right. I've never modified the `.sawcore` parts of this repo before, so let me see if I can figure this out. If not, I'll shout for help.

**Comment by RyanGlScott** [comment]
@robdockins: I've pushed a newer version that passes the `cryptol-saw-core` test suite. There are a handful of things remaining that I'm not sure about:

* This code special-cases `Literal`:

  https://github.com/GaloisInc/saw-core/blob/1dce64877bb720f48db9500ca1d399c6fc8a5e8e/cryptol-saw-core/src/Verifier/SAW/Cryptol.hs#L160-L167
  
  Should it also special-case `LiteralLessThan`?
* This code special-cases `PLiteral`:

  https://github.com/GaloisInc/saw-core/blob/1dce64877bb720f48db9500ca1d399c6fc8a5e8e/cryptol-saw-core/src/Verifier/SAW/Cryptol/Simpset.hs#L44
  
  Should it also special-case `PLiteralLessThan`?
* Now that I've changed `Cryptol.sawcore`, does this warrant regenerating the Coq code in `saw-core-coq`?

File: cryptol-saw-core/saw/Cryptol.sawcore
Comment (RyanGlScott) [inline cryptol-saw-core/saw/Cryptol.sawcore:949]: I'm not entirely sure about the use of `Nat` here, since the `n` in `LiteralLessThan n a` can be `inf`. Should this be `Num` instead?

Comment (robdockins) [reply cryptol-saw-core/saw/Cryptol.sawcore:949]: No, we'll only ever ask for literals of finite values, even if the upper bound is `inf` (NB, it's an exclusive upper bound!)

**Comment by robdockins** [comment]
I think @brianhuffman will have some better context on these questions, and I'll defer to him on this.

I think we have a choice to make: we can either directly reuse the same dictionaries for `PLiteralLessThan` and `PLiteral`, or they can be distinct.  If we make them the same, that makes our job slightly easier, but it is maybe confusing.

Regarding normalizeProp: we do probably need to do something there.  What we do depends on the question above.

Regarding saw-core-coq, I'm not sure.

Regardless of if we decide to merge the dictionaries for `PLiteral` and `PLiteralLessThan`, we need to make sure that the correct dictionaries get selected in `proveProp`.

**Comment by RyanGlScott** [comment]
@brianhuffman, have you had a chance to look at this?

**Comment by brianhuffman** [comment]
I think this is all good, but I'd like to extend the `cryptol-saw-core` test suite to make sure that the translator handles all the implicit class instance rules correctly.

**Comment by brianhuffman** [comment]
Extending the `instance.cry` test case has revealed a missing case in `proveProp`:
```
Translated Cryptol.cry!
Translated Float.cry!
Translated Array.cry!
Translated superclass.cry!
cryptol-saw-core-tc-test: You have encountered a bug in cryptol-saw-core's implementation.
*** Please create an issue at https://github.com/GaloisInc/saw-core/issues

%< ---------------------------------------------------
  Revision:  5b78eaebf0a48806e9970f1dcb5b81552b3b3bc2
  Branch:    PLiteralLessThan (uncommited files present)
  Location:  proveProp
  Message:   LiteralLessThan val`1085 [n`1086]
CallStack (from HasCallStack):
  panic, called at src/Verifier/SAW/Cryptol/Panic.hs:13:9 in cryptol-saw-core-0.1-C2qbXKeswyJ2OpS52uQUKk:Verifier.SAW.Cryptol.Panic
  panic, called at src/Verifier/SAW/Cryptol.hs:623:17 in cryptol-saw-core-0.1-C2qbXKeswyJ2OpS52uQUKk:Verifier.SAW.Cryptol
  proveProp, called at src/Verifier/SAW/Cryptol.hs:931:23 in cryptol-saw-core-0.1-C2qbXKeswyJ2OpS52uQUKk:Verifier.SAW.Cryptol
%< ---------------------------------------------------
```

**Comment by brianhuffman** [comment]
Note: While the OP says this PR adapts to GaloisInc/cryptol@ddbb673, the additions to the test suite technically require some later commits from GaloisInc/cryptol#1085.

**Review by brianhuffman — APPROVED** [review]
I think this is all ready, and the `cryptol-saw-core` test suite passes, but it might be good to let @robdockins have a look before merging.

**Review by robdockins — APPROVED** [review]
Looks about right to me.

We should probably backport those changes to the instances test into Cryptol.  I don't remember if I added them before.

**Comment by brianhuffman** [comment]
Those definitions in `instances.cry` are indeed not present in the cryptol test suite, and we should add them there.

**Comment by robdockins** [comment]
OK https://github.com/GaloisInc/cryptol/issues/1144

**Comment by RyanGlScott** [comment]
Thanks, @brianhuffman!

**Pull request merged**
